File: examples/Hybrid_City10000.cpp
/* ----------------------------------------------------------------------------

 * GTSAM Copyright 2010-2020, Georgia Tech Research Corporation,
 * Atlanta, Georgia 30332-0415
 * All Rights Reserved
 * Authors: Frank Dellaert, et al. (see THANKS for the full author list)

 * See LICENSE for the license information

 * -------------------------------------------------------------------------- */

/**
 * @file   Hybrid_City10000.cpp
 * @brief  Example of using hybrid estimation
 *         with multiple odometry measurements.
 * @author Varun Agrawal
 * @date   January 22, 2025
 */

#include <gtsam/geometry/Pose2.h>
#include <gtsam/hybrid/HybridNonlinearFactor.h>
#include <gtsam/hybrid/HybridNonlinearFactorGraph.h>
#include <gtsam/hybrid/HybridSmoother.h>
#include <gtsam/hybrid/HybridValues.h>
#include <gtsam/inference/Symbol.h>
#include <gtsam/nonlinear/Values.h>
#include <gtsam/slam/BetweenFactor.h>
#include <gtsam/slam/PriorFactor.h>
#include <gtsam/slam/dataset.h>
#include <time.h>

#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <fstream>
#include <string>
#include <vector>

using namespace std;
using namespace gtsam;
using namespace boost::algorithm;

using symbol_shorthand::L;
using symbol_shorthand::M;
using symbol_shorthand::X;

const size_t kMaxLoopCount = 2000;  // Example default value
const size_t kMaxNrHypotheses = 10;

auto kOpenLoopModel = noiseModel::Diagonal::Sigmas(Vector3::Ones() * 10);

auto kPriorNoiseModel = noiseModel::Diagonal::Sigmas(
    (Vector(3) << 0.0001, 0.0001, 0.0001).finished());

auto kPoseNoiseModel = noiseModel::Diagonal::Sigmas(
    (Vector(3) << 1.0 / 30.0, 1.0 / 30.0, 1.0 / 100.0).finished());

// Experiment Class
class Experiment {
 private:
  std::string filename_;
  HybridSmoother smoother_;
  HybridNonlinearFactorGraph graph_;
  Values initial_;
  Values result_;

  /**
   * @brief Write the result of optimization to file.
   *
   * @param result The Values object with the final result.
   * @param num_poses The number of poses to write to the file.
   * @param filename The file name to save the result to.
   */
  void writeResult(const Values& result, size_t numPoses,
                   const std::string& filename = "Hybrid_city10000.txt") {
    ofstream outfile;
    outfile.open(filename);

    for (size_t i = 0; i < numPoses; ++i) {
      Pose2 outPose = result.at<Pose2>(X(i));
      outfile << outPose.x() << " " << outPose.y() << " " << outPose.theta()
              << std::endl;
    }
    outfile.close();
    std::cout << "Output written to " << filename << std::endl;
  }

  /**
   * @brief Create a hybrid loop closure factor where
   * 0 - loose noise model and 1 - loop noise model.
   */
  HybridNonlinearFactor hybridLoopClosureFactor(size_t loopCounter, size_t keyS,
                                                size_t keyT,
                                                const Pose2& measurement) {
    DiscreteKey l(L(loopCounter), 2);

    auto f0 = std::make_shared<BetweenFactor<Pose2>>(
        X(keyS), X(keyT), measurement, kOpenLoopModel);
    auto f1 = std::make_shared<BetweenFactor<Pose2>>(
        X(keyS), X(keyT), measurement, kPoseNoiseModel);

    std::vector<NonlinearFactorValuePair> factors{
        {f0, kOpenLoopModel->negLogConstant()},
        {f1, kPoseNoiseModel->negLogConstant()}};
    HybridNonlinearFactor mixtureFactor(l, factors);
    return mixtureFactor;
  }

  /// @brief Create hybrid odometry factor with discrete measurement choices.
  HybridNonlinearFactor hybridOdometryFactor(
      size_t numMeasurements, size_t keyS, size_t keyT, const DiscreteKey& m,
      const std::vector<Pose2>& poseArray,
      const SharedNoiseModel& poseNoiseModel) {
    auto f0 = std::make_shared<BetweenFactor<Pose2>>(
        X(keyS), X(keyT), poseArray[0], poseNoiseModel);
    auto f1 = std::make_shared<BetweenFactor<Pose2>>(
        X(keyS), X(keyT), poseArray[1], poseNoiseModel);

    std::vector<NonlinearFactorValuePair> factors{{f0, 0.0}, {f1, 0.0}};
    HybridNonlinearFactor mixtureFactor(m, factors);
    return mixtureFactor;
  }

  /// @brief Perform smoother update and optimize the graph.
  void smootherUpdate(HybridSmoother& smoother,
                      HybridNonlinearFactorGraph& graph, const Values& initial,
                      size_t kMaxNrHypotheses, Values* result) {
    HybridGaussianFactorGraph linearized = *graph.linearize(initial);
    smoother.update(linearized, kMaxNrHypotheses);
    // throw if x0 not in hybridBayesNet_:
    const KeySet& keys = smoother.hybridBayesNet().keys();
    if (keys.find(X(0)) == keys.end()) {
      throw std::runtime_error("x0 not in hybridBayesNet_");
    }
    graph.resize(0);
    // HybridValues delta = smoother.hybridBayesNet().optimize();
    // result->insert_or_assign(initial.retract(delta.continuous()));
  }

 public:
  /// Construct with filename of experiment to run
  explicit Experiment(const std::string& filename)
      : filename_(filename), smoother_(0.99) {}

  /// @brief Run the main experiment with a given maxLoopCount.
  void run(size_t maxLoopCount) {
    // Prepare reading
    ifstream in(filename_);
    if (!in.is_open()) {
      cerr << "Failed to open file: " << filename_ << endl;
      return;
    }

    // Initialize local variables
    size_t discreteCount = 0, index = 0;
    size_t loopCount = 0;

    std::list<double> timeList;

    // Set up initial prior
    double x = 0.0;
    double y = 0.0;
    double rad = 0.0;

    Pose2 priorPose(x, y, rad);
    initial_.insert(X(0), priorPose);
    graph_.push_back(PriorFactor<Pose2>(X(0), priorPose, kPriorNoiseModel));

    // Initial update
    clock_t beforeUpdate = clock();
    smootherUpdate(smoother_, graph_, initial_, kMaxNrHypotheses, &result_);
    clock_t afterUpdate = clock();
    std::vector<std::pair<size_t, double>> smootherUpdateTimes;
    smootherUpdateTimes.push_back({index, afterUpdate - beforeUpdate});

    // Start main loop
    size_t keyS = 0, keyT = 0;
    clock_t startTime = clock();
    std::string line;
    while (getline(in, line) && index < maxLoopCount) {
      std::vector<std::string> parts;
      split(parts, line, is_any_of(" "));

      keyS = stoi(parts[1]);
      keyT = stoi(parts[3]);

      int numMeasurements = stoi(parts[5]);
      std::vector<Pose2> poseArray(numMeasurements);
      for (int i = 0; i < numMeasurements; ++i) {
        x = stod(parts[6 + 3 * i]);
        y = stod(parts[7 + 3 * i]);
        rad = stod(parts[8 + 3 * i]);
        poseArray[i] = Pose2(x, y, rad);
      }

      // Flag to decide whether to run smoother update
      bool doSmootherUpdate = false;

      // Take the first one as the initial estimate
      Pose2 odomPose = poseArray[0];
      if (keyS == keyT - 1) {
        // Odometry factor
        if (numMeasurements > 1) {
          // Add hybrid factor
          DiscreteKey m(M(discreteCount), numMeasurements);
          HybridNonlinearFactor mixtureFactor = hybridOdometryFactor(
              numMeasurements, keyS, keyT, m, poseArray, kPoseNoiseModel);
          graph_.push_back(mixtureFactor);
          discreteCount++;
          doSmootherUpdate = true;
          std::cout << "mixtureFactor: " << keyS << " " << keyT << std::endl;
        } else {
          graph_.add(BetweenFactor<Pose2>(X(keyS), X(keyT), odomPose,
                                          kPoseNoiseModel));
        }
        // Insert next pose initial guess
        initial_.insert(X(keyT), initial_.at<Pose2>(X(keyS)) * odomPose);
      } else {
        // Loop closure
        HybridNonlinearFactor loopFactor =
            hybridLoopClosureFactor(loopCount, keyS, keyT, odomPose);
        // print loop closure event keys:
        std::cout << "Loop closure: " << keyS << " " << keyT << std::endl;
        graph_.add(loopFactor);
        doSmootherUpdate = true;
        loopCount++;
      }

      if (doSmootherUpdate) {
        gttic_(SmootherUpdate);
        beforeUpdate = clock();
        smootherUpdate(smoother_, graph_, initial_, kMaxNrHypotheses, &result_);
        afterUpdate = clock();
        smootherUpdateTimes.push_back({index, afterUpdate - beforeUpdate});
        gttoc_(SmootherUpdate);
        doSmootherUpdate = false;
      }

      // Record timing for odometry edges only
      if (keyS == keyT - 1) {
        clock_t curTime = clock();
        timeList.push_back(curTime - startTime);
      }

      // Print some status every 100 steps
      if (index % 100 == 0) {
        std::cout << "Index: " << index << std::endl;
        if (!timeList.empty()) {
          std::cout << "Acc_time: " << timeList.back() / CLOCKS_PER_SEC
                    << " seconds" << std::endl;
          // delta.discrete().print("The Discrete Assignment");
          tictoc_finishedIteration_();
          tictoc_print_();
        }
      }

      index++;
    }

    // Final update
    beforeUpdate = clock();
    smootherUpdate(smoother_, graph_, initial_, kMaxNrHypotheses, &result_);
    afterUpdate = clock();
    smootherUpdateTimes.push_back({index, afterUpdate - beforeUpdate});

    // Final optimize
    gttic_(HybridSmootherOptimize);
    HybridValues delta = smoother_.optimize();
    gttoc_(HybridSmootherOptimize);

    result_.insert_or_assign(initial_.retract(delta.continuous()));

    std::cout << "Final error: " << smoother_.hybridBayesNet().error(delta)
              << std::endl;

    clock_t endTime = clock();
    clock_t totalTime = endTime - startTime;
    std::cout << "Total time: " << totalTime / CLOCKS_PER_SEC << " seconds"
              << std::endl;

    // Write results to file
    writeResult(result_, keyT + 1, "Hybrid_City10000.txt");

    // TODO Write to file
    //  for (size_t i = 0; i < smoother_update_times.size(); i++) {
    //    auto p = smoother_update_times.at(i);
    //    std::cout << p.first << ", " << p.second / CLOCKS_PER_SEC <<
    //    std::endl;
    //  }

    // Write timing info to file
    ofstream outfileTime;
    std::string timeFileName = "Hybrid_City10000_time.txt";
    outfileTime.open(timeFileName);
    for (auto accTime : timeList) {
      outfileTime << accTime << std::endl;
    }
    outfileTime.close();
    std::cout << "Output " << timeFileName << " file." << std::endl;
  }
};

/* ************************************************************************* */
int main() {
  Experiment experiment(findExampleDataFile("T1_city10000_04.txt"));
  // Experiment experiment("../data/mh_T1_city10000_04.txt"); //Type #1 only
  // Experiment experiment("../data/mh_T3b_city10000_10.txt"); //Type #3 only
  // Experiment experiment("../data/mh_T1_T3_city10000_04.txt"); //Type #1 +
  // Type #3

  // Run the experiment
  experiment.run(kMaxLoopCount);

  return 0;
}

File: python/gtsam/examples/HybridCity10000.py
"""
GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
Atlanta, Georgia 30332-0415
All Rights Reserved

See LICENSE for the license information

Script for running hybrid estimator on the City10000 dataset.

Author: Varun Agrawal
"""

import argparse
import time

import numpy as np
from gtsam.symbol_shorthand import L, M, X

import gtsam
from gtsam import (BetweenFactorPose2, HybridNonlinearFactor,
                   HybridNonlinearFactorGraph, HybridSmoother, HybridValues,
                   Pose2, PriorFactorPose2, Values)


def parse_arguments():
    """Parse command line arguments"""
    parser = argparse.ArgumentParser()
    parser.add_argument("--data_file",
                        help="The path to the City10000 data file",
                        default="T1_city10000_04.txt")
    return parser.parse_args()


# Noise models
open_loop_model = gtsam.noiseModel.Diagonal.Sigmas(np.ones(3) * 10)
open_loop_constant = open_loop_model.negLogConstant()

prior_noise_model = gtsam.noiseModel.Diagonal.Sigmas(
    np.asarray([0.0001, 0.0001, 0.0001]))

pose_noise_model = gtsam.noiseModel.Diagonal.Sigmas(
    np.asarray([1.0 / 30.0, 1.0 / 30.0, 1.0 / 100.0]))
pose_noise_constant = pose_noise_model.negLogConstant()


class City10000Dataset:
    """Class representing the City10000 dataset."""

    def __init__(self, filename):
        self.filename_ = filename
        try:
            self.f_ = open(self.filename_, 'r')
        except OSError:
            print(f"Failed to open file: {self.filename_}")

    def __del__(self):
        self.f_.close()

    def read_line(self, line: str, delimiter: str = " "):
        """Read a `line` from the dataset, separated by the `delimiter`."""
        return line.split(delimiter)

    def parse_line(self, line: str) -> tuple[list[Pose2], tuple[int, int]]:
        """Parse line from file"""
        parts = self.read_line(line)

        key_s = int(parts[1])
        key_t = int(parts[3])

        num_measurements = int(parts[5])
        pose_array = [Pose2()] * num_measurements

        for i in range(num_measurements):
            x = float(parts[6 + 3 * i])
            y = float(parts[7 + 3 * i])
            rad = float(parts[8 + 3 * i])
            pose_array[i] = Pose2(x, y, rad)

        return pose_array, (key_s, key_t)

    def next(self):
        """Read and parse the next line."""
        line = self.f_.readline()
        if line:
            return self.parse_line(line)
        else:
            return None, None


class Experiment:
    """Experiment Class"""

    def __init__(self,
                 filename: str,
                 marginal_threshold: float = 0.9999,
                 max_loop_count: int = 8000,
                 update_frequency: int = 3,
                 max_num_hypotheses: int = 10,
                 relinearization_frequency: int = 10):
        self.dataset_ = City10000Dataset(filename)
        self.max_loop_count = max_loop_count
        self.update_frequency = update_frequency
        self.max_num_hypotheses = max_num_hypotheses
        self.relinearization_frequency = relinearization_frequency

        self.smoother_ = HybridSmoother(marginal_threshold)
        self.new_factors_ = HybridNonlinearFactorGraph()
        self.all_factors_ = HybridNonlinearFactorGraph()
        self.initial_ = Values()

    def hybrid_loop_closure_factor(self, loop_counter, key_s, key_t,
                                   measurement: Pose2):
        """
        Create a hybrid loop closure factor where
        0 - loose noise model and 1 - loop noise model.
        """
        l = (L(loop_counter), 2)
        f0 = BetweenFactorPose2(X(key_s), X(key_t), measurement,
                                open_loop_model)
        f1 = BetweenFactorPose2(X(key_s), X(key_t), measurement,
                                pose_noise_model)
        factors = [(f0, open_loop_constant), (f1, pose_noise_constant)]
        mixture_factor = HybridNonlinearFactor(l, factors)
        return mixture_factor

    def hybrid_odometry_factor(self, key_s, key_t, m,
                               pose_array) -> HybridNonlinearFactor:
        """Create hybrid odometry factor with discrete measurement choices."""
        f0 = BetweenFactorPose2(X(key_s), X(key_t), pose_array[0],
                                pose_noise_model)
        f1 = BetweenFactorPose2(X(key_s), X(key_t), pose_array[1],
                                pose_noise_model)

        factors = [(f0, pose_noise_constant), (f1, pose_noise_constant)]
        mixture_factor = HybridNonlinearFactor(m, factors)

        return mixture_factor

    def smoother_update(self, max_num_hypotheses) -> float:
        """Perform smoother update and optimize the graph."""
        print(f"Smoother update: {self.new_factors_.size()}")
        before_update = time.time()
        linearized = self.new_factors_.linearize(self.initial_)
        self.smoother_.update(linearized, max_num_hypotheses)
        self.all_factors_.push_back(self.new_factors_)
        self.new_factors_.resize(0)
        after_update = time.time()
        return after_update - before_update

    def reInitialize(self) -> float:
        """Re-linearize, solve ALL, and re-initialize smoother."""
        print(f"================= Re-Initialize: {self.all_factors_.size()}")
        before_update = time.time()
        self.all_factors_ = self.all_factors_.restrict(
            self.smoother_.fixedValues())
        linearized = self.all_factors_.linearize(self.initial_)
        bayesNet = linearized.eliminateSequential()
        delta: HybridValues = bayesNet.optimize()
        self.initial_ = self.initial_.retract(delta.continuous())
        self.smoother_.reInitialize(bayesNet)
        after_update = time.time()
        print(f"Took {after_update - before_update} seconds.")
        return after_update - before_update

    def run(self):
        """Run the main experiment with a given max_loop_count."""
        # Initialize local variables
        discrete_count = 0
        index = 0
        loop_count = 0
        update_count = 0

        time_list = []  #list[(int, float)]

        # Set up initial prior
        priorPose = Pose2(0, 0, 0)
        self.initial_.insert(X(0), priorPose)
        self.new_factors_.push_back(
            PriorFactorPose2(X(0), priorPose, prior_noise_model))

        # Initial update
        update_time = self.smoother_update(self.max_num_hypotheses)
        smoother_update_times = []  # list[(int, float)]
        smoother_update_times.append((index, update_time))

        # Flag to decide whether to run smoother update
        number_of_hybrid_factors = 0

        # Start main loop
        result = Values()
        start_time = time.time()

        while index < self.max_loop_count:
            pose_array, keys = self.dataset_.next()
            if pose_array is None:
                break
            key_s = keys[0]
            key_t = keys[1]

            num_measurements = len(pose_array)

            # Take the first one as the initial estimate
            odom_pose = pose_array[0]
            if key_s == key_t - 1:
                # Odometry factor
                if num_measurements > 1:
                    # Add hybrid factor
                    m = (M(discrete_count), num_measurements)
                    mixture_factor = self.hybrid_odometry_factor(
                        key_s, key_t, m, pose_array)
                    self.new_factors_.push_back(mixture_factor)

                    discrete_count += 1
                    number_of_hybrid_factors += 1
                    print(f"mixture_factor: {key_s} {key_t}")
                else:
                    self.new_factors_.push_back(
                        BetweenFactorPose2(X(key_s), X(key_t), odom_pose,
                                           pose_noise_model))

                # Insert next pose initial guess
                self.initial_.insert(
                    X(key_t),
                    self.initial_.atPose2(X(key_s)) * odom_pose)
            else:
                # Loop closure
                loop_factor = self.hybrid_loop_closure_factor(
                    loop_count, key_s, key_t, odom_pose)

                # print loop closure event keys:
                print(f"Loop closure: {key_s} {key_t}")
                self.new_factors_.push_back(loop_factor)
                number_of_hybrid_factors += 1
                loop_count += 1

            if number_of_hybrid_factors >= self.update_frequency:
                update_time = self.smoother_update(self.max_num_hypotheses)
                smoother_update_times.append((index, update_time))
                number_of_hybrid_factors = 0
                update_count += 1

                if update_count % self.relinearization_frequency == 0:
                    self.reInitialize()

            #  Record timing for odometry edges only
            if key_s == key_t - 1:
                cur_time = time.time()
                time_list.append(cur_time - start_time)

            # Print some status every 100 steps
            if index % 100 == 0:
                print(f"Index: {index}")

                if len(time_list) != 0:
                    print(f"Accumulate time: {time_list[-1]} seconds")

            index += 1

        # Final update
        update_time = self.smoother_update(self.max_num_hypotheses)
        smoother_update_times.append((index, update_time))

        # Final optimize
        delta = self.smoother_.optimize()

        result.insert_or_assign(self.initial_.retract(delta.continuous()))

        print(f"Final error: {self.smoother_.hybridBayesNet().error(delta)}")

        end_time = time.time()
        total_time = end_time - start_time
        print(f"Total time: {total_time} seconds")

        # Write results to file
        self.write_result(result, key_t + 1, "Hybrid_City10000.txt")

        # Write timing info to file
        self.write_timing_info(time_list=time_list)

    def write_result(self, result, num_poses, filename="Hybrid_city10000.txt"):
        """
        Write the result of optimization to file.

        Args:
            result (Values): he Values object with the final result.
            num_poses (int): The number of poses to write to the file.
            filename (str): The file name to save the result to.
        """
        with open(filename, 'w') as outfile:

            for i in range(num_poses):
                out_pose = result.atPose2(X(i))
                outfile.write(
                    f"{out_pose.x()} {out_pose.y()} {out_pose.theta()}\n")

        print(f"Output written to {filename}")

    def write_timing_info(self,
                          time_list,
                          time_filename="Hybrid_City10000_time.txt"):
        """Log all the timing information to a file"""

        with open(time_filename, 'w') as out_file_time:

            for acc_time in time_list:
                out_file_time.write(f"{acc_time}\n")

            print(f"Output {time_filename} file.")


def main():
    """Main runner"""
    args = parse_arguments()

    experiment = Experiment(gtsam.findExampleDataFile(args.data_file))
    experiment.run()


if __name__ == "__main__":
    main()
